lib/Dialect/CGGI/Transforms/BooleanLineVectorizer.cpp

```diff
@@ -0,0 +1,224 @@
+    // Each loop will have the key and a bucket with all the operations in
+    for (auto &[key, bucket] : compatibleOps) {
+      if (bucket.size() < 2) {
+        continue;
+      }
+
+      LLVM_DEBUG({
+        llvm::dbgs() << "[START] Bucket \t Vectorizing ops:\n";
+        for (auto op : bucket) {
+          llvm::dbgs() << " - " << *op << "\n";
+        }
+      });
+
+      OpBuilder builder(bucket.back());
+      // relies on CGGI ops having a single result type
+      Type elementType = key->getResultTypes()[0];
+      RankedTensorType tensorType = RankedTensorType::get(
+          {static_cast<int64_t>(bucket.size())}, elementType);
+
+      SmallVector<Value, 4> vectorizedOperands;
+      SmallVector<StringAttr, 4> vectorizedGateOperands;
+
+      for (auto op : bucket) {
+        std::string str;
+        if (isa<cggi::AndOp>(op)) {
+          str = "and";
+        } else if (isa<cggi::NandOp>(op)) {
+          str = "nand";
+        } else if (isa<cggi::XorOp>(op)) {
+          str = "xor";
+        } else if (isa<cggi::XNorOp>(op)) {
+          str = "xnor";
+        } else if (isa<cggi::OrOp>(op)) {
+          str = "or";
+        } else if (isa<cggi::NorOp>(op)) {
+          str = "nor";
+        }
+        vectorizedGateOperands.push_back(StringAttr::get(&context, str));
+      }
+
+      // Group the independent operands over the operations
+      for (uint operandIndex = 0; operandIndex < key->getNumOperands();
+           ++operandIndex) {
+        SmallVector<Value, 4> operands;
+        LLVM_DEBUG({
+          llvm::dbgs() << "For: " << key->getName()
+                       << " Number of ops: " << key->getNumOperands() << "\n";
+        });
+
+        operands.reserve(bucket.size());
+        ///------------------------------------------
+        for (auto *op : bucket) {
+          LLVM_DEBUG(llvm::dbgs() << "getOperand for [" << operandIndex << "]: "
+                                  << op->getOperand(operandIndex) << "\n");
+          operands.push_back(op->getOperand(operandIndex));
+        }
+        ///------------------------------------------
+
+        auto fromElementsOp = builder.create<tensor::FromElementsOp>(
+            key->getLoc(), tensorType, operands);
+        vectorizedOperands.push_back(fromElementsOp.getResult());
+      }
+
+      LLVM_DEBUG({
+        llvm::dbgs() << "Go over vectorizedOps:\n";
+        for (auto op : vectorizedOperands) {
+          llvm::dbgs() << " - " << op << "\n";
+        }
+        llvm::dbgs() << "Go over vectorizedGateOps:\n";
+        for (auto op : vectorizedGateOperands) {
+          llvm::dbgs() << " - " << op;
+        }
+        llvm::dbgs() << "\n";
+      });
+
+      auto oplist = CGGIGateAttr::get(&context, vectorizedGateOperands);
+
+      auto vectorizedOp = builder.create<cggi::PackedOp>(
+          key->getLoc(), tensorType, oplist, vectorizedOperands[0],
+          vectorizedOperands[1]);
+
+      vectorizedOp->getResult(0).setType(tensorType);
```

Comment:
interesting! just curious, why was this needed? I would assume the first operand in the builder above was for the result type.